### PR TITLE
[Refactoring] ConvertConcatenationToInterpolatedString: Support inlineing of other interpolated strings

### LINKED ISSUE #48900: Support Interpolating Ternary Operations
**Brief description:**
It would be nice if Roslyn would offer to interpolate cases where a ternary operator has been slammed in the front (or back end) of a interpolated string.

**Languages applicable:**

My request would be C# only; but I can't see why it wouldn't work in VB.NET

**Code example that the analyzer should report:**
```csharp
(validateOnly ? "Validating" : "Sorting") + $" solution `{targetArgument}`";
```

Can be Converted to

```csharp
$"{(validateOnly ? "Validating" : "Sorting")} solution `{targetArgument}`"
```

**Additional information:**
The reason a user might have written the above is they've encountered an error when attempting to originally interpolate the string, not realizing that you can wrap it in `()`. See https://github.com/dotnet/roslyn/issues/21259 This stack overflow issue shows that there are at least a couple hundred users attempting to use this within an interpolated string https://stackoverflow.com/questions/31844058/how-to-use-the-ternary-operator-inside-an-interpolated-string

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

### PR DESCRIPTION
**Pull request by MaStr11**
Fixes #48900
Supersedes #49207

### CODE REVIEW

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 1]
```diff
@@ -154,6 +155,37 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
                         content.Add(generator.InterpolatedStringText(generator.InterpolatedStringTextToken(textWithoutQuotes)));
                     }
                 }
+                else if (syntaxFacts.IsInterpolatedStringExpression(piece) &&
+                    syntaxFacts.IsVerbatimInterpolatedStringExpression(piece) == isVerbatimStringLiteral)
+                {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:160]: please add a comment explaining the case this is matching, and what you're going to->from.

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 2]
```diff
@@ -154,6 +155,37 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
...
+                            }
+                            else
+                            {
+                                content.Add(contentPart);
+                            }
+                            previousContentWasStringLiteralExpression = true;
+                        }
+                        else
+                        {
+                            Contract.Fail();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:185]: either `throw ExceptionUtilities.UnexpectedValue()`; or `Debug.Fail`.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs [hunk 3]
```diff
@@ -2132,5 +2132,19 @@ public bool IsThrowStatement(SyntaxNode node)
...
+            var interpolatedStringExpression = (InterpolatedStringExpressionSyntax)node;
+            stringStartToken = interpolatedStringExpression.StringStartToken;
+            content = interpolatedStringExpression.Contents;
+            stringEndToken = interpolatedStringExpression.StringEndToken;
+        }
+
+        public bool IsVerbatimInterpolatedStringExpression(SyntaxNode node)
+        {
+            var interpolatedStringExpression = (InterpolatedStringExpressionSyntax)node;
+            return interpolatedStringExpression.StringStartToken.IsKind(SyntaxKind.InterpolatedVerbatimStringStartToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:2147]:
```suggestion
            return node is InterpolatedStringExpressionSyntax interpolatedString && interpolatedString.StringStartToken.IsKind(SyntaxKind.InterpolatedVerbatimStringStartToken);
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb [hunk 4]
```diff
@@ -2402,5 +2402,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.LanguageServices
...
+            Dim interpolatedStringExpressionSyntax As InterpolatedStringExpressionSyntax = DirectCast(node, InterpolatedStringExpressionSyntax)
+            stringStartToken = interpolatedStringExpressionSyntax.DollarSignDoubleQuoteToken
+            content = interpolatedStringExpressionSyntax.Contents
+            stringEndToken = interpolatedStringExpressionSyntax.DoubleQuoteToken
+        End Sub
+
+        Function IsVerbatimInterpolatedStringExpression(node As SyntaxNode) As Boolean Implements ISyntaxFacts.IsVerbatimInterpolatedStringExpression
+            Return False
+        End Function
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:2416]:
```suggestion
```

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs (hunk 1 above)
Comment (MaStr11) [reply src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:160]: Done in 95928ba

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs (hunk 2 above)
Comment (MaStr11) [reply src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs]: Done in 95928ba

**Comment by MaStr11** [comment]
@CyrusNajmabadi Thanks for the review. 

In my previous PR I added support for the removal of superfluous parentheses in `Interpolation.Expression`.  I used the existing `CanRemoveParentheses(SemanticModel semanticModel)` extension method. So in case there are potentially removable expressions in the new `InterploatedString`, I requested a new sematicModel and removed the parentheses if possible:

https://github.com/MaStr11/roslyn/blob/b7928515a2f32da7b73803d9ffbe7f452a5eec1d/src/Features/CSharp/Portable/CodeRefactorings/ConvertStringConcatToInterpolated/CSharpConvertStringConcatToInterpolatedRefactoringProvider.cs#L104-L144
```CS
        private static async Task<Document> UpdateDocumentAsync(Document document, BinaryExpressionSyntax binaryExpression, ImmutableArray<ExpressionSyntax> concatParts, CancellationToken cancellationToken)
        {
            var interpolatedStringContent = ConvertConcatPartsToInterpolatedStringContent(concatParts);
            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
            var editor = new SyntaxEditor(root, CSharpSyntaxGenerator.Instance);
            var interpolated = InterpolatedStringExpression(
                Token(SyntaxKind.InterpolatedStringStartToken),
                List(interpolatedStringContent));
            // expressions may have superfluous parenthesis after the conversion to an interpolated string:
            // "a" + (1 + 1) -> $"a{(1 + 1)}"
            // if there is any (InterpolationSyntax.Expression is ParenthesizedExpressionSyntax) we need to
            // recompile and check via parenthesized.CanRemoveParentheses(semanticModel) if it can be removed.
            var anyPotentiallyRemovableParenthesis = interpolated.Contents.Any(IsParenthesizedInterpolation);
            var annotation = new SyntaxAnnotation("InterpolatedStringExpressionWithPotentiallyRemovableParanthesisKind");
            if (anyPotentiallyRemovableParenthesis)
            {
                // annotate "interpolated" because we need to find it after the changes are applied
                interpolated = interpolated.WithAdditionalAnnotations(annotation);
            }
            editor.ReplaceNode(binaryExpression, interpolated);

            if (anyPotentiallyRemovableParenthesis)
            {
                // Take the changed document and go through interpolatedNode.Contents to check if any parenthesis can be removed.
                var newDocument = document.WithSyntaxRoot(editor.GetChangedRoot());
                var newRoot = await newDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
                var interpolatedNode = (InterpolatedStringExpressionSyntax)newRoot.GetAnnotatedNodes(annotation).Single();
                var semanticModel = await newDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
                editor = new SyntaxEditor(newRoot, CSharpSyntaxGenerator.Instance);
                foreach (var content in interpolatedNode.Contents.Where(IsParenthesizedInterpolation).Cast<InterpolationSyntax>())
                {
                    var parenthesized = (ParenthesizedExpressionSyntax)content.Expression;
                    if (parenthesized.CanRemoveParentheses(semanticModel))
                    {
                        editor.ReplaceNode(parenthesized, parenthesized.Expression);
                    }
                }
            }

            return document.WithSyntaxRoot(editor.GetChangedRoot());
        }
```

Do we want to reuse that approach here or not?

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 5]
```diff
@@ -166,29 +165,18 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
...
-                            {
-                                content.Add(contentPart);
-                            }
-                        }
-                        else if (syntaxFacts.IsInterpolation(contentPart))
-                        {
-                            content.Add(contentPart);
-                            previousContentWasStringLiteralExpression = false;
+                            // if piece starts with a text and the previous part was a string, merge the two parts (see also above)
+                            var newText = ConcatinateTextToTextNode(generator, content.Last(), contentPart.GetFirstToken().Text);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:171]: 1. spelling: "Concatinate"
2. consider giving an example of what this means.

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 6]
```diff
@@ -156,6 +156,29 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
...
+                        {
+                            // if piece starts with a text and the previous part was a string, merge the two parts (see also above)
+                            var newText = ConcatinateTextToTextNode(generator, content.Last(), contentPart.GetFirstToken().Text);
+                            content[^1] = newText;
+                        }
+                        else
+                        {
+                            content.Add(contentPart);
+                        }
+                        // Only the first contentPart can be merged, therefore we set previousContentWasStringLiteralExpression to false
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:178]:
```suggestion
                        }
                        
                        // Only the first contentPart can be merged, therefore we set previousContentWasStringLiteralExpression to false
```

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs (hunk 5 above)
Comment (MaStr11) [reply src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs]: Done in 6c75e923 and 2772cef7

**Review by CyrusNajmabadi — APPROVED** [review]
Thanks!

**Comment by MaStr11** [comment]
@CyrusNajmabadi Sorry, but I haven't seen your LGTM before I pushed more commits. The commits are only minor clean-ups. I have one more question though: Should I add some logic to remove superfluous parenthesis? There are some approaches to do that. I implemented one with tests as described here: https://github.com/dotnet/roslyn/pull/49229#issuecomment-724012609. I could do this as part of this PR or a new one.

**Comment by CyrusNajmabadi** [comment]
>  I have one more question though: Should I add some logic to remove superfluous parenthesis? 

In general, an easy way to do that is to just add the SimplifierAnnotation to the topmost node you're creating.  We should then go through and clean up those parens.  See if that works for you.  Otherwise, i wouldn't bother doing more here (unless it was super easy).

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Sorry, but I haven't seen your LGTM before I pushed more commits.

that's fine.  i still review the later commits :)

**Comment by MaStr11** [comment]
> SimplifierAnnotation to the topmost node

The Simplifier worked like a charm, but I needed to place it on each `Expression` of a newly generated `InterploationSyntax` (I also tried to place it on the `InterpolatedStringExpression` but that didn't work).

**Comment by MaStr11** [comment]
@CyrusNajmabadi Tests are green now. I added the simplifier annotation to the expressions that get capsulated in an Interpolation, but only if the expression is a ParenthesisExpression.

**Comment by aolszowka** [comment]
I am sorry I am new to this process, where does this go? How does this get mainlined?

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**